Remove unnecessary uses of MD2 in X.509 parsing test suite

hanno-arm opened the pull request:
As has been reported in #821, many test cases in the X.509 parsing test suite use MD2 in places where an arbitrary hash algorithm would be supported. This is bad for two reasons: Firstly, it shows bad practice using a weak digest, and secondly, it significantly reduces negative test coverage in the default configuration which doesn't have MD2 enabled.

This PR changes the tests to use SHA-256 in place of MD2 in those instances where a general hash algorithm is allowed (and we're not specifically testing support for MD2).

hanno-arm commented on tests/suites/test_suite_x509parse.data:
```diff
@@ -12,7 +12,7 @@ x509_cert_info:"data_files/test-ca.crt":"cert. version     \: 3\nserial number
 X509 Certificate information MD2 Digest
 depends_on:MBEDTLS_PEM_PARSE_C:MBEDTLS_RSA_C:MBEDTLS_MD2_C
-x509_cert_info:"data_files/cert_md2.crt":"cert. version     \: 3\nserial number     \: 09\nissuer name       \: C=NL, O=PolarSSL, CN=PolarSSL Test CA\nsubject name      \: C=NL, O=PolarSSL, CN=PolarSSL Cert MD2\nissued  on        \: 2009-07-12 10\:56\:59\nexpires on        \: 2011-07-12 10\:56\:59\nsigned using      \: RSA with MD2\nRSA key size      \: 2048 bits\nbasic constraints \: CA=false\n"
+x509_cert_info:"data_files/cert_md2.crt":"cert. version     \: 3\nserial number     \: 09\nissuer name       \: C=NL, O=PolarSSL, CN=PolarSSL Test CA\nsubject name      \: C=NL, O=PolarSSL, CN=PolarSSL Cert MD2\nissued  on        \: 2009-07-12 10\:56\:59\nexpires on        \: 2011-07-12 10\:56\:59\nsigned using      \: RSA with SHA-256\nRSA key size      \: 2048 bits\nbasic constraints \: CA=fals ...
```
@mpg No, I'm sorry, that was careless.

hanno-arm commented:
@mpg @AndrzejKurek I need to rework this PR, my approach to replacing MD2 by SHA-256 was too careless. Please don't review this for now.

hanno-arm commented:
@mpg Could you take another look?

hanno-arm commented:
Nothing has changed since @AndrzejKurek's approval - I think it got lost when @mpg re-requested it. Was there a particular reason for that, @mpg?

mpg commented:
I don't remember doing that, so probably no particular reason ^^

hanno-arm commented:
Ping @Patater for gatekeeping.

mpg commented:
This fixes #821 which is labeled as a bug, so relabeling this with "bug", and hence needing backports.

mpg commented:
(Demoting from "ready for merge" as it needs backports.)

hanno-arm commented:
@k-stachowiak @Patater @AndrzejKurek I had to force-push this to resolve a ChangeLog conflict with `development`. Can you please re-review?

hanno-arm commented on ChangeLog:
```diff
@@ -132,6 +132,7 @@ Changes
      been disabled for lack of a sufficiently recent version of GnuTLS on the CI.
    * Ciphersuites based on 3DES now have the lowest priority by default when
      they are enabled.
+   * Replace multiple uses of MD2 by SHA-256 in X.509 test suite. Fixes #821.
```
Thanks for spotting, fixed it.

hanno-arm commented:
@Patater @AndrzejKurek I fixed the ChangeLog issue Jaeden spotted, and also rebased on top of `development`. Could you please re-review?

Patater commented on tests/suites/test_suite_x509parse.data:
```diff
@@ -1821,47 +1821,47 @@ mbedtls_x509_csr_parse:"30173014020100300D310B3009060355040613024E4C300100":"":M
 X509 CSR ASN.1 (bad attributes: missing)
 depends_on:MBEDTLS_ECDSA_C:MBEDTLS_ECP_DP_SECP256R1_ENABLED
-mbedtls_x509_csr_parse:"3081973081940201003034310B3009060355040613024E4C3111300F060355040A1308506F6C617253534C31123010060355040313096C6F63616C686F73743059301306072A8648CE3D020106082A8648CE3D0301070342000437CC56D976091E5A723EC7592DFF206EEE7CF9069174D0AD14B5F768225962924EE500D82311FFEA2FD2345D5D16BD8A88C26B770D55CD8A2A0EFA01C8B4EDFF":"":MBEDTLS_ERR_X509_INVALID_FORMAT + MBEDTLS_ERR_ASN1_OUT_OF_DATA
+mbedtls_x509_csr_parse:"3081973081940201003034310b3009060355040613024e4c3111300f060355040a1308506f6c617253534c31123010060355040313096c6f63616c686f73743059301306072a8648ce3d020106082a8648ce3d0301070342000437cc56d976091e5a723ec7592dff206eee7cf9069174d0ad14b5f768225962924ee500d82311ffea2fd2345d5d16bd8a88c26b770d55cd8a2a0efa01c8b4edff":"":MBEDTLS_ERR_X509_INVALID_FORMAT + MBEDTLS_ERR_ASN1_OUT_OF_DATA
```
Some new tests added since your original PR was raised use UPPERCASE. The commit "Consistently use lower case hex data in X.509 parsing tests" should update these tests as well, if it's to do what it says on the tin.

hanno-arm replied:
Right, thanks for spotting - actually, there were quite a number of tests missing this conversion.

hanno-arm replied:
This is addressed in 32b48ea.

Patater commented on tests/suites/test_suite_x509parse.data:
```diff
@@ -354,7 +354,7 @@ X509 CSR Information RSA-PSS with SHA512
 mbedtls_x509_csr_info:"data_files/server9.req.sha512":"CSR version   \: 1\nsubject name  \: C=NL, O=PolarSSL, CN=localhost\nsigned using  \: RSASSA-PSS (SHA512, MGF1-SHA512, 0x3E)\nRSA key size  \: 1024 bits\n"
 
-X509 CSR Information RSA with SHA256 - Microsoft header
+X509 CSR Information RSA with SHA-256 - Microsoft header
```
In commit "Use SHA-256 instead of MD2 in X.509 CRT parsing tests ", some tests still use MD2, like:

```
X509 Certificate ASN1 (TBSCertificate v3, ext CertificatePolicies tag, bool len missing)
depends_on:MBEDTLS_RSA_C:MBEDTLS_MD2_C
```

hanno-arm replied:
Right, I'll fix it.

hanno-arm replied:
This is addressed in ba698da.

AndrzejKurek commented:
Please ping me when the remaining issues are resolved @hanno-arm.

hanno-arm commented:
Thanks @Patater for spotting the inaccuracies, I believe to have addressed them. Could you please re-review?

Ping @AndrzejKurek for re-review, too!

hanno-arm commented:
@gilles-peskine-arm Thanks for your review.

The PR doesn't make the situation worse, because it only replaces MD2 by SHA-256 in parsing tests, and a MD2-specific parsing test remains ("X509 Certification information MD2 digest").

Nonetheless, you have a point that there's a pre-existing gap in the tests in that we don't exercise verification of an MD2 signed certificate, and I've added an analogue of "Valid Cert MD4 Digest" for MD2.

Also, I noticed that these tests don't expect success because they use a profile which forbids MDx, and I've added three tests which use a test-only profile allowing all MDs, so that we now also cover successful verification of MDx-signed CRTs.

@gilles-peskine-arm @AndrzejKurek Can you please re-review?

gilles-peskine-arm commented on include/mbedtls/config.h:
```diff
@@ -2511,7 +2511,7 @@
  *
  */
-//#define MBEDTLS_MD2_C
+#define MBEDTLS_MD2_C
```
You accidentally committed a modified `config.h`.

hanno-arm replied:
Damn... sorry

hanno-arm replied:
Fixed.
